Extension/src/LanguageServer/copilotCompletionContextProvider.ts

```diff
@@ -0,0 +1,185 @@
+    private async waitForCompletionWithTimeoutAndCancellation<T>(promise: Promise<T>, defaultValue: T | undefined,
+        timeout: number, token: vscode.CancellationToken): Promise<[T | undefined, SnippetsKind]> {
+        const defaultValuePromise = new Promise<T>((resolve, reject) => setTimeout(() => {
+            if (token.isCancellationRequested) {
+                reject(new CancellationError());
+            } else {
+                reject(new DefaultValueFallback());
+            }
+        }, timeout));
+        const cancellationPromise = new Promise<T>((_, reject) => {
+            token.onCancellationRequested(() => {
+                reject(new CancellationError());
+            });
+        });
+        let snippetsOrNothing: T | undefined;
+        try {
+            snippetsOrNothing = await Promise.race([promise, cancellationPromise, defaultValuePromise]);
+        } catch (e) {
+            if (e instanceof DefaultValueFallback) {
+                return [defaultValue, defaultValue !== undefined ? SnippetsKind.GotFromCache : SnippetsKind.MissingCacheMiss];
+            } else if (e instanceof CancellationError) {
+                return [undefined, SnippetsKind.Cancelled];
+            } else {
+                throw e;
+            }
+        }
+
+        return [snippetsOrNothing, SnippetsKind.Computed];
+    }
+
+    // Get the completion context with a timeout and a cancellation token.
+    // The cancellationToken indicates that the value should not be returned nor cached.
+    private async getCompletionContextWithCancellation(documentUri: string, caretOffset: number,
+        startTime: number, out: Logger, telemetry: CopilotContextTelemetry, token: vscode.CancellationToken): Promise<CodeSnippet[]> {
+        try {
+            const docUri = vscode.Uri.parse(documentUri);
+            const snippets = await clients.getClientFor(docUri).getCompletionContext(docUri, caretOffset, token);
+
+            const codeSnippets = snippets.context.map((item) => {
+                if (token.isCancellationRequested) {
+                    telemetry.addCancelledLate();
+                    throw new CancellationError();
+                }
+                return {
+                    importance: item.importance, uri: item.uri, value: item.text
+                };
+            });
+
+            this.completionContextCache.set(documentUri, codeSnippets);
+            const duration: number = performance.now() - startTime;
+            out.appendLine(`Copilot: getCompletionContextWithCancellation(): Cached in [ms]: ${duration}`);
+            telemetry.addSnippetCount(codeSnippets?.length);
+            telemetry.addCacheComputedElapsed(duration);
+
+            return codeSnippets;
+        } catch (e) {
+            const err = e as Error;
+            out.appendLine(`Copilot: getCompletionContextWithCancellation(): Error: '${err?.message}', stack '${err?.stack}`);

```

Comment:
It doesn't seem like we should output an error message on Cancel, since Cancelling seems like a non-error. Also, it seems like telemetry would want that tracked differently too.